[READY!] Adds alien queen sprite toggle

PKPenguin321 opened the pull request:
Fixes https://github.com/tgstation/tgstation/issues/16335

Adds an action button that toggles your sprite into the old small alien queen sprite so you can actually interact with things that are above you (and then toggle it back on later to feel stronk)
Supposedly you'll only appear small to yourself and not to anybody else but I can't confirm yet because this does not work

Todo:
- [x] Make it work at all ~~(fix the runtimes)~~
- [x] Make the action button have a purplish alien background instead of generic greyish 
- [x] Finalize the button (fix the [pixel shift error](http://puu.sh/qM1nY/ef8893214a.png))

:cl: PKPenguin321
add: Xeno queens now have an action button that makes them appear to themselves as the old, small queen sprite. Now they should be able to interact with things that are above them more easily.
/:cl:

PKPenguin321 committed:
shit i forgot i was on master branch

PKPenguin321 committed:
adds alien queen sprite toggle

RemieRichards commented on code/modules/mob/living/carbon/alien/humanoid/queen.dm:
```diff
@@ -22,7 +22,20 @@
+	..()
+	if(!small)
+		var/image/I = image('icons/mob/alien.dmi', "alienq_running")
+		owner.add_alt_appearance("alienq_running", I, list(src))
```
because when I wrote the example for you, ```src``` was the queen, as it needs to be, you've since moved it to an action toggle (which is a good idea well done) so src is no longer the queen, the queen will be the ```owner``` var.

also this key doesn't need to be ```"alienq_running"``` here, and I'd suggest something like ```"small_queen"```

PKPenguin321 replied:
ah alright cool

RemieRichards commented:
>Make the action button have a purplish alien background instead of generic greyish blue

that's just ```background_icon_state = "bg_alien"``` on the action

PKPenguin321 commented:
okay awesome thanks for the help :smile: 

also this is a fix

RemieRichards commented:
@PKPenguin321 it's not a fix, it's a workaround feature.

PKPenguin321 commented:
¯\\\_(ツ)_/¯

PKPenguin321 committed:
fixes
- makes the background of the button the right color
- fixes the runtime (but the button still doesnt work properly)

PKPenguin321 commented:
okay weird
@RemieRichards i updated it to do what you said. the runtime is gone and when i click the button i can see through VV that the "small" var is updating properly meaning the button is working, but the alien queens sprite does not change at all, which means something's still up :frowning: send help

RemieRichards commented:
are you the queen? because it only displays for the queen, of course.

PKPenguin321 commented:
yes, i spawned a queen then took direct control, both via admin commands

PKPenguin321 commented:
ack what conflicts how

RemieRichards commented:
the actions.dmi file, probably.

RemieRichards commented:
Aha, @PKPenguin321 set ```I.override``` to 1, I forgot it in the original example.

PKPenguin321 commented:
wtf i think git is freaking out it says that the conflicting file is `code/modules/food_and_drinks/recipes/tablecraft/recipes_burger.dm`

PKPenguin321 committed:
Merge branch 'master' of https://github.com/PKPenguin321/-tg-station into alien-queen-giga-toggle

# Conflicts:
#	code/modules/food_and_drinks/recipes/tablecraft/recipes_burger.dm (????????)
also actions.dmi i dont know why the fuck the burgers conflicted

PKPenguin321 commented:
jesus christ what

PKPenguin321 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into alien-queen-giga-toggle

# Conflicts:
#	icons/mob/actions.dmi
for real this time

PKPenguin321 commented:
thank you lord

PKPenguin321 committed:
aaaaaaa

PKPenguin321 commented:
@RemieRichards now it's throwing a runtime again!!! http://pastebin.com/2xRfPgBw

RemieRichards commented:
@PKPenguin321 because you're not passing an image anymore, you're passing the result of ```I.override = 1``` (which is a number, which is then used as an index for the internal AltAppearance lists, which is causing your runtime)
you DON'T do assignments like that inside proc calls.

PKPenguin321 committed:
bjkajkjbks

PKPenguin321 commented:
now the runtime is gone but the sprite still doesn't change

WJohn commented:
The old queen sprite is so ugly, I never updated the 1x1 queen's sprite to the new xenos at all and I don't really plan to. Would it be possible instead to make the button make you transparent and unable to click on yourself isntead?

PKPenguin321 commented:
wj it's fine, nobody even sees the tiny sprite but the queen itself, and even then it's optional for the queen

it's just more playable this way

Bawhoppen commented:
I want it so when you hover over it like makes your disappear but then your outline border is still there but apparently thats very costly

RemieRichards commented:
@AnturK I completely forgot, do we have to cleanup action datums manually for mobs? or is it done somewhere super high up the tree (like /atom, or /mob)?

RemieRichards commented:
also @WJohn no, images, whether overlays or overrides cannot have different ```mouse_opacity``` to their parent.

PKPenguin321 commented:
btw @RemieRichards  dont know if you caught word or if you're already working on it but just in case, ill let you know again:

the same issue before is happening, there is no runtime and the button works fine (the "small" var is toggling on and off properly) but the new appearance is never actually applied to the xeno queen. nobody in coderbus knew why this happened and i was hoping you did. i'm guessing we're using the proc wrong?

RemieRichards commented:
@PKPenguin321 the ```image()``` call needs ```loc = owner``` as an argument, the image isn't showing up because it's in nullspace.

Kromgar commented:
Can we please have ash drake button to make them appear smaller too? I dont know you could replace their sprite with ash drake armor or something. Ash drakes are near unplayable due to their massive wing size and height. Attacking anyway. 

I dunno not a lot of sprites to use for an ash drake. I suppose it could be requested

GeneralGoose commented:
make it replace sprite with a letter
ASCII meme

PKPenguin321 committed:
finally makes the button work

PKPenguin321 commented:
latest commit finally makes the button work! however, the xeno retains a pixel shift to the left that i need to remove when it shrinks and reapply when it grows
in this image you see that the xeno is half in the window. it only appears that way because of the shift
![ack](http://puu.sh/qM1nY/ef8893214a.png)
if anybody knows how to fix this let me know!

RemieRichards commented:
>Finalize the button (fix the pixel shift error)

you need to pixel shift the image (```I```) the opposite of the queen's current pixel offsets.
```
I.pixel_x = -owner.pixel_x
I.pixel_y = -owner.pixel_y
```
That'll undo the centering that was necessary for the large icon

PKPenguin321 commented:
It's perfect!
Thanks so much for the help remie :grin:

PKPenguin321 committed:
hopefully final commit

Kromgar commented:
@WJohn just make a new 1x1 queen then if your so worried.

This exists because people keep biting themselves when they try and attack people in melee

PKPenguin321 commented:
Should be good to merge!
